### PR TITLE
Better profiling experience

### PR DESCRIPTION
**Pull request by ttuegel**
This pull request improves the profiling experience.

You must use `cabal` to build the package, and you must have version 3.0 or later (run `cabal --version` to check). Stack does not seem to allow the user to override the profiling options passed to GHC, but this is essential for acceptable performance. You can install `cabal` and GHC with [`ghcup`](https://www.haskell.org/ghcup/).

To build:

```
# This writes the cabal.project.local file.
# The setting persists until you --disable-profiling or delete the file.
cabal configure --enable-profiling
cabal build kore
# At the end of the build, cabal will print the path of
# the kore-exec binary it built, you need to copy it into place:
cp [...]/kore-exec ./.build/kore/bin
```

With these changes, profiling overhead is about 500%, which is better than the 3000% we were seeing with Stack.

I am adding appropriate `INLINE` pragmas to get reasonable profiling reports. (Anything marked `INLINE` will not be reported automatically.)

Fixes: #1346 
See also: #1335 

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/src/Kore/Step/Strategy.hs [hunk 1]
```diff
@@ -291,14 +296,17 @@ insEdge (fromNode, toNode) exe@ExecutionGraph { graph } =
...
+    :: (NFData config, NFData rule)
+    => MonadState (Gr config (Seq rule)) m
     => ChildNode config rule
     -> m Graph.Node
-insChildNode configNode =
-    State.state insChildNodeWorker
+insChildNode childNode = do
+    graph <- deepseq childNode State.get
+    let (node', graph') = insChildNodeWorker graph
+    deepseq graph' (State.put graph')
```
Comment (ana-pantilie) [inline kore/src/Kore/Step/Strategy.hs]: Just to be sure I understand, this is what you were saying in the meeting about fully evaluating the graph when adding a new node, similar to calling `graph` after each step?

File: kore/src/Kore/ModelChecker/Bounded.hs [hunk 2]
```diff
@@ -61,6 +61,7 @@ import Kore.Step.Strategy
     , pickFinal
     , runStrategyWithSearchOrder
     )
+import Kore.Strategies.ProofState ()
```
Comment (ana-pantilie) [inline kore/src/Kore/ModelChecker/Bounded.hs]: Why was this added?

**Comment by ttuegel** [comment]
My workaround for the issue makes the KEVM tests time-out, so I must have affected time performance, too. :slightly_frowning_face:

**Comment by ttuegel** [comment]
I moved all the mitigations to #1378 so that we can merge this pull request, which now only adds the profiling options.
